src/ucp/rndv/proto_rndv.inl

```diff
@@ -265,6 +265,82 @@ ucp_proto_rndv_bulk_max_payload(ucp_request_t *req,
     return max_payload;
 }
 
+static UCS_F_ALWAYS_INLINE size_t
+ucp_proto_rndv_align_frag(size_t max_payload, size_t min_frag,
+                          size_t max_frag)
+{
+    ucs_assert(max_frag > min_frag);
+
+    if (max_payload < min_frag) {
+        return ucs_align_up(ucs_max(min_frag, max_payload),
+                            UCP_PROTO_RNDV_ALIGN);
+    }
+    return ucs_min(ucs_align_down(max_frag, UCP_PROTO_RNDV_ALIGN),
+                   ucs_align_up(max_payload, UCP_PROTO_RNDV_ALIGN));
+}
+
+static UCS_F_ALWAYS_INLINE size_t
+ucp_proto_rndv_align_first_frag(ucp_request_t *req, void   *buffer,
+                                size_t min_frag, size_t max_frag,
+                                size_t max_payload)
+{
+    size_t remain  = (size_t)buffer % UCP_PROTO_RNDV_ALIGN;
+
+    if (0 == remain) {
+        return ucp_proto_rndv_align_frag(max_payload, min_frag, max_frag);
+    }
+    req->flags |= UCP_REQUEST_FLAG_RNDV_ALIGN_FRAG;
+    return ucs_align_up((min_frag + remain), UCP_PROTO_RNDV_ALIGN) - remain;
+}
+
+static UCS_F_ALWAYS_INLINE size_t
+ucp_proto_rndv_adjust_align_next_frag(ucp_request_t *req,
+                                      const ucp_proto_rndv_bulk_priv_t *rpriv,
+                                      const ucp_proto_multi_lane_priv_t *lpriv,
+                                      const size_t max_payload)
+{
+    size_t min_frag      = rpriv->mpriv.min_frag;
+    size_t max_frag      = lpriv->max_frag;
+    size_t total_offset  = req->send.rndv.offset +
+                           req->send.state.dt_iter.offset;
+    size_t total_length  = ucp_proto_rndv_request_total_length(req);
+    size_t align_frag, remain_length;
+    void   *buffer;
+    unsigned is_max_frag = (total_length >= rpriv->mpriv.max_frag_sum);
+    unsigned buffer_is_aligned;
+
+    ucs_assert(req->send.state.dt_iter.dt_class == UCP_DATATYPE_CONTIG);
+
+    buffer = (void*)((size_t)req->send.state.dt_iter.type.contig.buffer +
+                     total_offset);
+    buffer_is_aligned = !(((size_t)buffer) % UCP_PROTO_RNDV_ALIGN);
+
+    if ((!total_offset && !is_max_frag) || (is_max_frag && !buffer_is_aligned)) {
+        if (ucs_unlikely(total_length <= (2 * UCP_PROTO_RNDV_ALIGN))) {
+            return total_length;
+        }
+        goto first_frag;
+    }
+    if (ucs_unlikely(is_max_frag && buffer_is_aligned)) {
+        return max_payload;
+    }
+
+    ucs_assert(total_length >= total_offset);
+    ucs_assert(total_length > (2 * UCP_PROTO_RNDV_ALIGN));
+    ucs_assert(1 == buffer_is_aligned);
```

Comment:
with bool vars no need to check for exact value
why is it expected to bye aligned when still need to do something?

Reply:
It is expected to be aligned because this code can be only executed if:
1. Each lane sends less than its maximal fragment size
2. Processing fragment with non-zero offset
It means that we should have already sent a fragment aligning address of the current fragment.

And the current fragment must have a size that does not violate the alignment of the next fragment.